### PR TITLE
Improve multi-threaded performance

### PR DESCRIPTION
**Pull request by hcho3**
* Use columnwise accessor to accelerate `ApplySplit()` step, with support for a compressed representation
* Parallel sort for evaluation step
* Cache gradient pairs when building histograms in `BuildHist()`

TODO: write a blog post

### CODE REVIEW

File: src/common/hist_util.cc [hunk 1]
```diff
@@ -147,20 +140,28 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat) {
         if (it == cend) it = cend - 1;
         unsigned idx = static_cast<unsigned>(it - cut->cut.begin());
         index[ibegin + j] = idx;
+        hit_count_tloc_[tid * nbins + idx]++;
```
Comment (tqchen) [inline src/common/hist_util.cc]: always a good habit to do ++x instead of x++

**Comment by tqchen** [comment]
@hcho3 can you rebase against the latest master and retrigger travis again? I have upgraded the testing requirement to be g++4.8 or higher which means full c++11 support

File: src/common/row_set.h [hunk 2]
```diff
@@ -21,11 +21,13 @@ class RowSetCollection {
   struct Elem {
     const bst_uint* begin;
     const bst_uint* end;
+    int node_id;
```
Comment (tqchen) [inline src/common/row_set.h]: comment on what -1 mean for node_id

File: src/tree/param.h [hunk 3]
```diff
@@ -111,6 +112,12 @@ struct TrainParam : public dmlc::Parameter<TrainParam> {
             "Tree growing policy. 0: favor splitting at nodes closest to the node, "
             "i.e. grow depth-wise. 1: favor splitting at nodes with highest loss "
             "change. (cf. LightGBM)");
+    DMLC_DECLARE_FIELD(colmat_dtype)
+        .set_default(static_cast<int>(DataType::kUInt32))
+        .add_enum("kUInt8", static_cast<int>(DataType::kUInt8))
```
Comment (tqchen) [inline src/tree/param.h]: use ```uint8```, similar for the rest part

Comment (tqchen) [inline src/tree/param.h]: is this something that user suppose to set?

File: src/common/column_matrix.h [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    unsigned row_ind_end;
+  };
+
+  std::vector<bst_uint> feature_counts_;
+  std::vector<ColumnType> type_;
+  std::vector<uint32_t> index_;  // index_: may store smaller integers; needs padding
+  std::vector<uint32_t> row_ind_;
+  std::vector<ColumnBoundary> boundary_;
+
+  size_t packing_factor_;
```
Comment (tqchen) [inline src/common/column_matrix.h]: add a comment about packing factor

File: src/tree/param.h (hunk 3 above)
Comment (hcho3) [inline src/tree/param.h]: Not usually. This is reserved for advanced use case.

Comment (tqchen) [inline src/tree/param.h]: i see, add a description then saying that this is reserved for advanced use

**Pull request merged**